tztrout/data.py

```diff
@@ -26,6 +26,67 @@
 AU_AREA_CODE_TO_STATE_MAP_PATH =  os.path.join(basepath, 'data/au_area_code_to_state.json')
 
 
+def deduplicator():
+    """Deduplicate strings in memory using a canonical mapping.
+
+    Works similarly to intern() but supports unicode as well.
+
+    >>> a = chr(1) + chr(1)  # Create two strings
+    >>> b = chr(1) + chr(1)
+
+    >>> a == b  # These have the same contents
+    True
+    >>> id(a) == id(b)  # But they're different objects
+    False
+
+    >>> deduplicate = deduplicator()  # Deduplicating them returns the same object
+    >>> id(deduplicate(a)) == id(deduplicate(b))
+    True
+    """
+    memo = {}
+
+    def deduplicate(item):
+        if item in memo:
+            return memo[item]
+        memo[item] = item
+        return item
+
+    return deduplicate
+
+
+class InMemoryZipData(object):
+    def __init__(self):
+
+        self.by_state = {}
+        self.by_city = {}
+        self.by_state_city = {}
+        deduplicate = deduplicator()
+
+        for zip in map(ZipCode, ZipCodeDatabase().conn_manager.query("SELECT * FROM ZipCodes", [])):
```

Comment:
Could use a more readable `ZipCodeDatabase().find_zip()` instead:
```
In [16]: zc = db.find_zip()[0]

In [17]: zc2 = pyzipcode.ZipCode(db.conn_manager.query("SELECT * FROM ZipCodes", [])[0])

In [18]: zc.city
Out[18]: u'Portsmouth'

In [19]: zc2.city
Out[19]: u'Portsmouth'

In [20]: zc.__dict__
Out[20]:
{'city': u'Portsmouth',
 'dst': 1,
 'latitude': 43.005895,
 'longitude': -71.013202,
 'state': u'NH',
 'timezone': -5,
 'zip': u'00210'}

In [21]: zc2.__dict__
Out[21]:
{'city': u'Portsmouth',
 'dst': 1,
 'latitude': 43.005895,
 'longitude': -71.013202,
 'state': u'NH',
 'timezone': -5,
 'zip': u'00210'}

In [22]:
```

Reply:
`format_result` used in ZipCodeDatabase builds an intermediate list of all zipcodes. I wanted to avoid that as building it is just wasted CPU time at startup.

Reply:
Ah ok. How about adding something like:
```
# XXX using map(...) instead of ZipCodeDatabase().find_zip() because the
# latter builds an intermediate list of all zipcodes. Having performance
# in mind, we want to avoid that.
```